### PR TITLE
Transition to .NETCore 3.0 for BuildXL

### PR DESCRIPTION
**Pull request by Kr1sso**
After the Nuget refactoring we are now able to introduce `.NETCoreApp3.+` and `.NETStandard2.+` frameworks to the BuildXL codebase. This PR bumps the version of .NETCore to 3.0 and allows us to begin shifting our entire release and build pipeline towards one target framework only.

### CODE REVIEW

**Comment by dannyvv** [comment]
Thank you for doing this!!!!

File: config.dsc [hunk 1]
```diff
@@ -78,7 +78,7 @@ config({
 
                 // Roslyn
                 { id: "Microsoft.Net.Compilers", version: "3.0.0" },
-                { id: "Microsoft.NETCore.Compilers", version: "2.10.0" },
+                { id: "Microsoft.NETCore.Compilers", version: "3.1.0-beta3-final" },
```
Comment (dannyvv) [inline config.dsc]: >3.1.0 [](start = 63, length = 5)

Title says 3.0... perhaps 3.1-beta3?

File: Public/Sdk/Public/Managed/helpers.dsc [hunk 2]
```diff
@@ -132,4 +136,44 @@ namespace Helpers {
...
         return results.toArray();
     };
+
+    /** TODO: Unofortunately the System.Interactive.Async package exposes IAsyncEnumerable in its own System.Generic.Collection namespace,
+    /*        colliding with .NETCore 3.0 implementation of the same name. For the time being (hopefully once the package maintainers updated their deps),
+    /*        we explicitly create an aliased reference for the code that depends on this.
+    **/
+
+    @@public
+    export function aliasedReferencesForSystemInteractiveAsync(references: Shared.Reference[]) : Csc.Arguments {
```
Comment (dannyvv) [inline Public/Sdk/Public/Managed/helpers.dsc]: >aliasedReferencesForSystemInteractiveAsync [](start = 20, length = 42)

Nit: Rename to indicate it is a hack to patch the referenes: i.e. `patchReferencesFor...` or `hackReferencesFor...`

File: Public/Sdk/Public/Managed/helpers.dsc [hunk 3]
```diff
@@ -132,4 +136,44 @@ namespace Helpers {
...
+        const needsSystemInteractiveAsync =
+            references.some(ref => (ref as Shared.ManagedNugetPackage).name !== undefined &&
```
Comment (dannyvv) [inline Public/Sdk/Public/Managed/helpers.dsc]: >(ref as Shared.ManagedNugetPackage [](start = 35, length = 34)

I beleive we have helper conditions:
`Shared.isManagedPackage(ref)` which will do the type coecion for you so you can write:
`Shared.isManagedPackage(ref) && ref.name.contains("....")`

so you don't have to do tostring either becaues nuget packages are always string for name.

File: Public/Sdk/Public/Managed/Frameworks/frameworks.dsc [hunk 4]
```diff
@@ -15,10 +15,10 @@ export const framework : Shared.Framework = (() => {
             return importFrom("Sdk.Managed.Frameworks.Net461").framework;
         case "net472":
             return importFrom("Sdk.Managed.Frameworks.Net472").framework;
-        case "netcoreapp2.2":
-            return importFrom("Sdk.Managed.Frameworks.NetCoreApp2.2").framework;
```
Comment (dannyvv) [inline Public/Sdk/Public/Managed/Frameworks/frameworks.dsc]: For completeness, please keep netcoreapp2.2

File: Public/Sdk/Public/Managed/Frameworks/netstandard2/module.config.dsc [hunk 5]
```diff
@@ -2,7 +2,7 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 module({
-    name: "Sdk.Managed.Frameworks.NetStandard2.0",
+    name: "Sdk.Managed.Frameworks.NetStandard2",
```
Comment (dannyvv) [inline Public/Sdk/Public/Managed/Frameworks/netstandard2/module.config.dsc]: May I ask why this change? The TFM is still NetStandard2.0. I was hoping to keep the module name to match the TFM

File: Public/Sdk/Public/Managed/Shared/frameworks.dsc [hunk 6]
```diff
@@ -55,7 +55,7 @@ namespace TargetFrameworks {
     export type DesktopTargetFrameworks = "net451" | "net461" | "net472";
 
     @@public
-    export type CoreClrTargetFrameworks = "netcoreapp2.2";
+    export type CoreClrTargetFrameworks = "netcoreapp3.0";
```
Comment (dannyvv) [inline Public/Sdk/Public/Managed/Shared/frameworks.dsc]: Same comment as before: Can we keep netcoreapp2.2? For completeness so that others can keep using it if they need to stay on 2.2?

File: Public/Sdk/Public/Managed/Testing/XUnit/xunit.dsc [hunk 7]
```diff
@@ -38,28 +38,28 @@ export function runConsoleTest(args: TestRunArguments): Result {
...
+        warningRegex: "^(?=a)b", // This is a never matching warning regex. StartOfLine followed by the next char must be 'a' (look ahead), and the next char must be a 'b'.
         workingDirectory: testDeployment.contents.root,
         retryExitCodes: Environment.getFlag("RetryXunitTests") ? [1] : [],
         unsafe: args.untrackTestDirectory ? {untrackedScopes: [testDeployment.contents.root]} : undefined,
     };
 
-    if (qualifier.targetFramework === "netcoreapp2.2") {
-        execArguments = importFrom("Sdk.Managed.Frameworks.NetCoreApp2.2").wrapInDotNetExeForCurrentOs(execArguments);
+    if (qualifier.targetFramework === "netcoreapp3.0") {
+        execArguments = importFrom("Sdk.Managed.Frameworks.NetCoreApp3").wrapInDotNetExeForCurrentOs(execArguments);
```
Comment (dannyvv) [inline Public/Sdk/Public/Managed/Testing/XUnit/xunit.dsc]: again to keep parity for when 3.1 will come out, why not have the .0 already to match the TFM?

File: Public/Sdk/SelfHost/BuildXL/StandaloneTest/StandaloneTest.dsc [hunk 8]
```diff
@@ -62,7 +62,7 @@ namespace StandaloneTest {
             `    description: "${assemblyName} [${qualifier.configuration}, ${dotNetFramework}]",`,
             `    testAssembly: f\`${def.assembly.name}\`,`,
             `    untracked: ${def.untracked},`,
-            `    wrapInDotNet: ${qualifier.targetFramework === "netcoreapp2.2"},`,
```
Comment (dannyvv) [inline Public/Sdk/SelfHost/BuildXL/StandaloneTest/StandaloneTest.dsc]: >qualifier.targetFramework === "netcoreapp2.2 [](start = 33, length = 44)

Should this be changed to `BuildXLSdk.isDotNetCoreBuild` or perhaps a new property `BuildXLSdk.IsNetCoreApp`?

File: Public/Src/Cache/ContentStore/Interfaces/BuildXL.Cache.ContentStore.Interfaces.dsc [hunk 9]
```diff
@@ -32,5 +32,6 @@ namespace Interfaces {
             "BuildXL.Cache.ContentStore.Distributed",
             "BuildXL.Cache.ContentStore.Interfaces.Test",
         ],
+        //tools : qualifier.targetFramework !== "netcoreapp3.0" ? undefined : NetCoreApp.aliasedSystemInteractiveAsyncAssembly
```
Comment (dannyvv) [inline Public/Src/Cache/ContentStore/Interfaces/BuildXL.Cache.ContentStore.Interfaces.dsc]: >//tools : qualifie [](start = 8, length = 18)

Remove?

File: Public/Src/Demos/Deployment.dsc [hunk 10]
```diff
@@ -14,7 +14,7 @@ namespace Deployment {
         ]
     };
 
-    const frameworkSpecificPart = qualifier.targetFramework === "netcoreapp2.2"
+    const frameworkSpecificPart = qualifier.targetFramework === "netcoreapp3.0"
```
Comment (dannyvv) [inline Public/Src/Demos/Deployment.dsc]: >ier.targetFramework === "netcoreapp [](start = 40, length = 35)

`BuildXLSdk.isNetCoreApp` or `!BuildXLSdk.isFullFramework`

Comment (dannyvv) [reply Public/Src/Demos/Deployment.dsc]: Note, will not comment on each and every one :)
Note not asking you to fix before merging but if you don't fix it let me know as I want to fix it :)

---
In reply to: [283563371](https://github.com/microsoft/BuildXL/pull/298#discussion_r283563371) [](ancestors = 283563371)

File: Public/Src/Deployment/cache.NugetPackages.dsc [hunk 11]
```diff
@@ -11,12 +11,12 @@ namespace Cache.NugetPackages {
...
+    const WinX64ContentStore = importFrom("BuildXL.Cache.ContentStore").withQualifier({ configuration: qualifier.configuration, targetFramework: "netcoreapp3.0", targetRuntime: "win-x64" });
 
     const Net451MemoizationStore = importFrom("BuildXL.Cache.MemoizationStore").withQualifier({ configuration: qualifier.configuration, targetFramework: "net451", targetRuntime: "win-x64" });
     const Net461MemoizationStore = importFrom("BuildXL.Cache.MemoizationStore").withQualifier({ configuration: qualifier.configuration, targetFramework: "net461", targetRuntime: "win-x64" });
     const Net472MemoizationStore = importFrom("BuildXL.Cache.MemoizationStore").withQualifier({ configuration: qualifier.configuration, targetFramework: "net472", targetRuntime: "win-x64" });
-    const WinX64MemoizationStore = importFrom("BuildXL.Cache.MemoizationStore").withQualifier({ configuration: qualifier.configuration, targetFramework: "netcoreapp2.2", targetRuntime: "win-x64" });
+    const WinX64MemoizationStore = importFrom("BuildXL.Cache.MemoizationStore").withQualifier({ configuration: qualifier.configuration, targetFramework: "netcoreapp3.0", targetRuntime: "win-x64" });
```
Comment (dannyvv) [inline Public/Src/Deployment/cache.NugetPackages.dsc]: >ildXL.Cache.MemoizationSt [](start = 49, length = 25)

I don't think AzDev is consuming our netcoreapp2.2 yet but please check with @kewili to be 100% sure since this might break them.

File: Public/Sdk/Public/Managed/Frameworks/netstandard2/module.config.dsc (hunk 5 above)
Comment (Kr1sso) [reply Public/Sdk/Public/Managed/Frameworks/netstandard2/module.config.dsc]: I thought it doesn't really matter what minor we use, hence I decided to call it `NetCoreApp3` and `NetStandard2`. If we wan't to keep more flavors it makes sense to be specific, I'll revert.

File: Public/Sdk/Public/Managed/Shared/frameworks.dsc (hunk 6 above)
Comment (Kr1sso) [reply Public/Sdk/Public/Managed/Shared/frameworks.dsc]: I'll add it back, but this brings back the runtime packages for `netcoreapp2.2` as aliased Nuget references and thus adds more "load" to the initial download process. Also, our self deployment qualifiers are `net472` and `netcoreapp3.0` now - can you elaborate on who 'they' are and what scenario that would entail?

File: Public/Sdk/Public/Managed/Testing/XUnit/xunit.dsc (hunk 7 above)
Comment (Kr1sso) [reply Public/Sdk/Public/Managed/Testing/XUnit/xunit.dsc]: Adjusted, see above.

File: Public/Src/Cache/ContentStore/Interfaces/BuildXL.Cache.ContentStore.Interfaces.dsc (hunk 9 above)
Comment (Kr1sso) [reply Public/Src/Cache/ContentStore/Interfaces/BuildXL.Cache.ContentStore.Interfaces.dsc]: Thanks, missed that one!

File: Public/Src/Demos/Deployment.dsc (hunk 10 above)
Comment (Kr1sso) [reply Public/Src/Demos/Deployment.dsc]: Fixed.

File: Public/Sdk/Public/Managed/helpers.dsc [hunk 12]
```diff
@@ -132,4 +136,43 @@ namespace Helpers {
...
+
+    /** TODO: Unofortunately the System.Interactive.Async package exposes IAsyncEnumerable in its own System.Generic.Collection namespace,
+    /*        colliding with .NETCore 3.0 implementation of the same name. For the time being (hopefully once the package maintainers updated their deps),
+    /*        we explicitly create an aliased reference for the code that depends on this.
+    **/
+
+    @@public
+    export function patchReferencesForSystemInteractiveAsync(references: Shared.Reference[]) : Csc.Arguments {
+        const needsSystemInteractiveAsync = references.some(ref =>
+            Shared.isManagedPackage(ref) && !Shared.isAssembly(ref) && ref.name.contains("System.Interactive.Async"));
```
Comment (dannyvv) [inline Public/Sdk/Public/Managed/helpers.dsc]: >!Shared.isAssembly(r [](start = 44, length = 20)

Curious why you needed this check? the first one should be exclusive...

File: Public/Sdk/Public/Managed/Shared/frameworks.dsc (hunk 6 above)
Comment (dannyvv) [reply Public/Sdk/Public/Managed/Shared/frameworks.dsc]: note the download only happens when used. Our nuget is still eager but when that one is lazy when not used it should just be a short web request.

---
In reply to: [283756439](https://github.com/microsoft/BuildXL/pull/298#discussion_r283756439) [](ancestors = 283756439)

Comment (dannyvv) [reply Public/Sdk/Public/Managed/Shared/frameworks.dsc]: future customers of DScript

---
In reply to: [283913281](https://github.com/microsoft/BuildXL/pull/298#discussion_r283913281) [](ancestors = 283913281,283756439)

File: Public/Sdk/Public/Managed/Frameworks/netstandard2/module.config.dsc (hunk 5 above)
Comment (dannyvv) [reply Public/Sdk/Public/Managed/Frameworks/netstandard2/module.config.dsc]: We should match what nguet does since they are the 'owners' of these names and the semantics in co-owership with .net team.

---
In reply to: [283755776](https://github.com/microsoft/BuildXL/pull/298#discussion_r283755776) [](ancestors = 283755776)

File: Public/Sdk/Public/Managed/helpers.dsc (hunk 12 above)
Comment (Kr1sso) [reply Public/Sdk/Public/Managed/helpers.dsc]: My best guess is we have precedence / ternary evaluation issues in DScript - as it did not work.
